File: USRP/Signals/lib/pulse_shaping.py
import numpy as np
import matplotlib.pyplot as plt
from scipy import signal

# Define the pulse shaping module (Rectangular or Root-raised-cosine filters are used in the Tx and Rx side).
# (split the filters to TX & RX side so the receiver filter could be used to attenuate out-of-band channel noise)
# Source code for the commpy filters invoked here could be found in the following link:
# https://pydoc.net/scikit-commpy/0.3.0/commpy.filters/


# The pulse shaping module takes the following arguments as inputs:

# a:                        Base band symbols (containing both the preamble and payload). 

# M:                        This is the transmitter-side oversampling factor, each symbol (bit) being transmitted is represented
#                           by M samples

# fs:                       This is the sampling rate of the DAC

# pulse_shape:              "rect" (rectangular) or "rrc" (root-raised cosine)

# alpha:                    This is the roll-off factor for RRC (valid value [0,1])

# L:                        This is the length (span) of the pulse-shaping filter (in the unit of symbols)

# The pulse shaping module returns the following argument as output:

# baseband                   This is the baseband signal

#filter length: make it a parameter w/ default value
#TODO: introduce memory and make pulse shaping streaming
def pulse_shaping(a, M, fs, pulse_shape, alpha, L):
        
        #Upsample by a factor of M
        y = signal.upfirdn([1],a,M)

        if(pulse_shape == 'rrc'):

                #Root Raised-Cosine span
                N = L*M

                T_symbol = 1/(fs/M)

                ##Square root raised-cosine (SRRC)

                time, h = rrcosfilter(N,alpha, T_symbol, fs)

                baseband = np.convolve(a,h)


        if(pulse_shape == 'rect'):

                Ts = 1/fs

                #rectangular pulse
                # h = np.ones(M)

                baseband = y


        return baseband


def rrcosfilter(N, alpha, Ts, Fs):
    """
    Generates a root raised cosine (RRC) filter (FIR) impulse response.
 
    Parameters
    ----------
    N : int
        Length of the filter in samples.
 
    alpha : float
        Roll off factor (Valid values are [0, 1]).
 
    Ts : float
        Symbol period in seconds.
 
    Fs : float
        Sampling Rate in Hz.
 
    Returns
    ---------
 
    h_rrc : 1-D ndarray of floats
        Impulse response of the root raised cosine filter.
 
    time_idx : 1-D ndarray of floats
        Array containing the time indices, in seconds, for
        the impulse response.
    """
 
    T_delta = 1/float(Fs)
    time_idx = ((np.arange(N)-N/2))*T_delta
    sample_num = np.arange(N)
    h_rrc = np.zeros(N, dtype=float)
 
    for x in sample_num:
        t = (x-N/2)*T_delta
        if t == 0.0:
            h_rrc[x] = 1.0 - alpha + (4*alpha/np.pi)
        elif alpha != 0 and t == Ts/(4*alpha):
            h_rrc[x] = (alpha/np.sqrt(2))*(((1+2/np.pi)* \
                    (np.sin(np.pi/(4*alpha)))) + ((1-2/np.pi)*(np.cos(np.pi/(4*alpha)))))
        elif alpha != 0 and t == -Ts/(4*alpha):
            h_rrc[x] = (alpha/np.sqrt(2))*(((1+2/np.pi)* \
                    (np.sin(np.pi/(4*alpha)))) + ((1-2/np.pi)*(np.cos(np.pi/(4*alpha)))))
        else:
            h_rrc[x] = (np.sin(np.pi*t*(1-alpha)/Ts) +  \
                    4*alpha*(t/Ts)*np.cos(np.pi*t*(1+alpha)/Ts))/ \
                    (np.pi*t*(1-(4*alpha*t/Ts)*(4*alpha*t/Ts))/Ts)
 
    return time_idx, h_rrc

File: SDR/tests/lib/IrishsatSignalProcessing/lib/src/pulse_shaping.c
//pulse_shaping.c

#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include "../standardArray.h"

// Function prototypes
Array_Tuple pulse_shaping(Array_Tuple a, int M, double fs, char pulse_shape[], double alpha, int L);
Array_Tuple rrcosfilter(int N, double alpha, double Ts, double Fs);

// save dynamic arrays into a static variable before freeing the array 
Array_Tuple pulse_shaping(Array_Tuple a, int M, double fs, char pulse_shape[], double alpha, int L) {
    if (strcmp(pulse_shape, "rrc") == 0) {
        int N = L * M;
        // Root Raised-Cosine span
        double T_symbol = 1.0 / (fs / M);

        // Generate RRC filter coefficients
        Array_Tuple h_rrc = rrcosfilter(N, alpha, T_symbol, fs);

        // Convolve the input with the RRC filter
        Array_Tuple a_imaj = zerosArray(a.length);
        Complex_Array_Tuple a_complex = {a, a_imaj};
        Complex_Array_Tuple convolve_out = convolve(a_complex, h_rrc);
        freeArrayMemory(a_imaj);
        freeArrayMemory(h_rrc);
        freeArrayMemory(convolve_out.imaginary);
        return convolve_out.real;
        //free(h_rrc);
    } 
    // else if (strcmp(pulse_shape, "rect") == 0) {//! This is not correct, goes off of scipy.upfirdn
    //     // Rectangular pulse
    //   for (int i = 0; i < a.length * M; i++) {
    //       baseband[i] = a.array[i];
    //   }
    //   Array_Tuple tuple = zerosArray(a.length * M);
    //   for (int i = 0; i < tuple.length; i++)
    //   {
    //     tuple.array[i] = baseband[i];
    //   }
    //   //free(baseband);
    //   return tuple;
    // }
    return a;
}

Array_Tuple rrcosfilter(int N, double alpha, double Ts, double Fs) {
    // Generate a root raised cosine (RRC) filter (FIR) impulse response
    double T_delta = 1.0 / Fs;

    // Allocate memory for time indices
    Array_Tuple time_idx_arange = arange(0, N, 1);
    Array_Tuple time_idx__minus = subtractDoubleFromArray(time_idx_arange, (double)N /2.0);
    Array_Tuple time_idx = multiplyDoubleFromArray(time_idx__minus, T_delta);

    Array_Tuple h_rrc = zerosArray(N);
    // Populate time indices
    for (int x = 0; x < N; x++) {
        time_idx.array[x] = (x - N / 2) * T_delta;
    }
    freeArrayMemory(time_idx_arange);
    freeArrayMemory(time_idx__minus);
    freeArrayMemory(time_idx);

    // Populate RRC filter coefficients
    for (int x = 0; x < N; x++) {
        double t = (x - (double)N / 2.0) * T_delta;
        if (t == 0.0) {
            h_rrc.array[x] = 1.0 - alpha + (4.0 * alpha / M_PI);
        } else if (alpha != 0 && t == Ts / (4 * alpha)) {
            h_rrc.array[x] = (alpha / sqrt(2)) * (((1 + 2.0 / M_PI) * (sin(M_PI / (4 * alpha)))) + ((1 - 2.0 / M_PI) * (cos(M_PI / (4 * alpha)))));
        } else if (alpha != 0 && t == -Ts / (4 * alpha)) {
            h_rrc.array[x] = (alpha / sqrt(2)) * (((1 + 2.0 / M_PI) * (sin(M_PI / (4 * alpha)))) + ((1 - 2.0 / M_PI) * (cos(M_PI / (4 * alpha)))));
        } else {
            h_rrc.array[x] = (sin(M_PI * t * (1 - alpha) / Ts) + 4 * alpha * (t / Ts) * cos(M_PI * t * (1 + alpha) / Ts)) / (M_PI * t * (1 - (4 * alpha * t / Ts) * (4 * alpha * t / Ts)) / Ts);
        }
    }

    return h_rrc;
}
